File: src/main/kotlin/magicbook/gtlitecore/api/gui/widget/SlotLikeButtonWidget.java
package magicbook.gtlitecore.api.gui.widget;

import com.cleanroommc.modularui.api.ITheme;
import com.cleanroommc.modularui.api.drawable.IDrawable;
import com.cleanroommc.modularui.drawable.DynamicDrawable;
import com.cleanroommc.modularui.drawable.GuiDraw;
import com.cleanroommc.modularui.drawable.ItemDrawable;
import com.cleanroommc.modularui.integration.jei.JeiIngredientProvider;
import com.cleanroommc.modularui.screen.viewport.ModularGuiContext;
import com.cleanroommc.modularui.theme.WidgetSlotTheme;
import com.cleanroommc.modularui.theme.WidgetTheme;
import com.cleanroommc.modularui.widgets.ButtonWidget;
import gregtech.api.items.metaitem.MetaItem;
import net.minecraft.client.renderer.GlStateManager;
import net.minecraft.item.ItemStack;
import org.jetbrains.annotations.Nullable;

import java.util.function.Supplier;

/**
 * Button widget that looks like and partially behaves like item slot. Draws supplied {@code ItemStack},
 * draws white overlay when hovered and allows interaction with JEI.
 */
public class SlotLikeButtonWidget extends ButtonWidget<SlotLikeButtonWidget> implements JeiIngredientProvider
{

    private final Supplier<ItemStack> itemSupplier;
    private final IDrawable itemDrawable;

    /**
     * @param itemStack {@code ItemStack} to "put into the slot".
     */
    public SlotLikeButtonWidget(ItemStack itemStack)
    {
        this.itemSupplier = () -> itemStack;
        this.itemDrawable = new ItemDrawable(itemStack).asIcon();
        disableHoverBackground();
    }

    public SlotLikeButtonWidget(MetaItem<?>.MetaValueItem metaItem)
    {
        this.itemSupplier = metaItem::getStackForm;
        this.itemDrawable = new ItemDrawable(metaItem.getStackForm()).asIcon();
        disableHoverBackground();
    }

    /**
     * @param itemSupplier  Supplier that specifies {@code ItemStack} to be present in the slot.
     */
    public SlotLikeButtonWidget(Supplier<ItemStack> itemSupplier)
    {
        this.itemSupplier = itemSupplier;
        this.itemDrawable = new DynamicDrawable(() -> new ItemDrawable(itemSupplier.get())).asIcon();
        disableHoverBackground();
    }

    @Override
    public WidgetTheme getWidgetThemeInternal(ITheme theme)
    {
        return theme.getItemSlotTheme();
    }

    @Override
    public void draw(ModularGuiContext context, WidgetTheme widgetTheme)
    {
        super.draw(context, widgetTheme);
        this.itemDrawable.drawAtZero(context, getArea(), widgetTheme);
        if (isHovering())
        {
            GlStateManager.disableLighting();
            GlStateManager.enableBlend();
            GlStateManager.colorMask(true, true, true, false);
            GuiDraw.drawRect(1, 1, 16, 16, getHoverColor(widgetTheme));
            GlStateManager.colorMask(true, true, true, true);
            GlStateManager.disableBlend();
        }
    }

    private int getHoverColor(WidgetTheme widgetTheme)
    {
        if (widgetTheme instanceof WidgetSlotTheme)
        {
            WidgetSlotTheme slotTheme = (WidgetSlotTheme) widgetTheme;
            return slotTheme.getSlotHoverColor();
        }
        return ITheme.getDefault()
                .getItemSlotTheme()
                .getSlotHoverColor();
    }

    @Nullable
    @Override
    public Object getIngredient()
    {
        return itemSupplier.get();
    }

}

File: src/main/kotlin/gregtechlite/gtlitecore/api/gui/widget/SlotLikeButtonWidget.kt
package gregtechlite.gtlitecore.api.gui.widget

import com.cleanroommc.modularui.api.ITheme
import com.cleanroommc.modularui.api.drawable.IDrawable
import com.cleanroommc.modularui.drawable.DynamicDrawable
import com.cleanroommc.modularui.drawable.GuiDraw
import com.cleanroommc.modularui.drawable.ItemDrawable
import com.cleanroommc.modularui.integration.jei.JeiIngredientProvider
import com.cleanroommc.modularui.screen.viewport.ModularGuiContext
import com.cleanroommc.modularui.theme.WidgetSlotTheme
import com.cleanroommc.modularui.theme.WidgetTheme
import com.cleanroommc.modularui.widgets.ButtonWidget
import gregtech.api.items.metaitem.MetaItem
import net.minecraft.client.renderer.GlStateManager
import net.minecraft.item.ItemStack

/**
 * Button widget that looks like and partially behaves like item slot.
 * Draws supplied [ItemStack], draws white overlay when hovered and allows interaction with JEI.
 */
class SlotLikeButtonWidget : ButtonWidget<SlotLikeButtonWidget>, JeiIngredientProvider
{

    private val itemSupplier: () -> ItemStack
    private val itemDrawable: IDrawable

    /**
     * @param itemStack The [ItemStack] to "put into the slot".
     */
    constructor(itemStack: ItemStack)
    {
        this.itemSupplier = { itemStack }
        this.itemDrawable = ItemDrawable(itemStack).asIcon()
        this.disableHoverBackground()
    }

    /**
     * @param metaItem The [MetaItem] to specify the correspondenced [ItemStack] to put into the slot.
     */
    constructor(metaItem: MetaItem<*>.MetaValueItem)
    {
        this.itemSupplier = { metaItem.stackForm }
        this.itemDrawable = ItemDrawable(metaItem.stackForm).asIcon()
        this.disableHoverBackground()
    }

    /**
     * @param itemSupplier Supplier that specifies `ItemStack` to be present in the slot.
     */
    constructor(itemSupplier: () -> ItemStack)
    {
        this.itemSupplier = itemSupplier
        this.itemDrawable = DynamicDrawable { ItemDrawable(itemSupplier.invoke()) }.asIcon()
        this.disableHoverBackground()
    }

    override fun getWidgetThemeInternal(theme: ITheme): WidgetTheme?
    {
        return theme.itemSlotTheme
    }

    override fun draw(context: ModularGuiContext?, widgetTheme: WidgetTheme?)
    {
        super.draw(context, widgetTheme)
        this.itemDrawable.drawAtZero(context, area, widgetTheme)
        if (this.isHovering)
        {
            GlStateManager.disableLighting()
            GlStateManager.enableBlend()
            GlStateManager.colorMask(true, true, true, false)
            GuiDraw.drawRect(1f, 1f, 16f, 16f, getHoverColor(widgetTheme))
            GlStateManager.colorMask(true, true, true, true)
            GlStateManager.disableBlend()
        }
    }

    private fun getHoverColor(widgetTheme: WidgetTheme?): Int
    {
        if (widgetTheme is WidgetSlotTheme)
        {
            val slotTheme = widgetTheme
            return slotTheme.slotHoverColor
        }
        return ITheme.getDefault().itemSlotTheme.slotHoverColor
    }

    override fun getIngredient(): Any? = itemSupplier.invoke()

}
